File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/java/DealnnoyNum.java
import java.io.*;
import java.lang.*;
import java.util.*;
import java.math.*;


class DealnnoyNum {
    
    /*@ public normal_behavior
      @ ensures \result == ((x % 4294967296L + 6442450944L) % 4294967296L) - 2147483648L;
      @ model public static pure \bigint fb_wrap(\bigint x);
      @*/
    /*@ public normal_behavior
      @ requires n == 0 || m == 0 || (n > 0 && m > 0);
      @ ensures \result == (n == 0 || m == 0 ? 1 : fb_delannoy(m - 1, n) + fb_delannoy(m - 1, n - 1) + fb_delannoy(m, n - 1));
      @ measured_by (n > 0 && m > 0 ? n + m : 0);
      @ model public static pure \bigint fb_delannoy(\bigint n, \bigint m);
      @*/
    /*@ public normal_behavior
      @ requires n == 0 || m == 0 || (n > 0 && m > 0);
      @ assignable \nothing;
      @ ensures \result == fb_wrap(fb_delannoy(n, m));
      @ measured_by (n > 0 && m > 0 ? (\bigint)n + m : 0);
      @*/
    //@ code_java_math
    public static int dealnnoyNum(int n, int m) {
        if (m == 0 || n == 0) {
            return 1;
        }
        return dealnnoyNum(m - 1, n) + dealnnoyNum(m - 1, n - 1) + dealnnoyNum(m, n - 1);
    }
}

File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/c/DealnnoyNum.c
#include "java_arrays.h"
#include <stdint.h>

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left - right);
*/
static int32_t java_sub(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left - (uint32_t)right);
}

/*@
  assigns \nothing;
  ensures \result == (int32_t)((integer)left + right);
*/
static int32_t java_add(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left + (uint32_t)right);
}

/*@
  logic integer fb_delannoy(integer n, integer m) =
    n <= 0 || m <= 0 ? 1 : fb_delannoy(m - 1, n) + fb_delannoy(m - 1, n - 1) + fb_delannoy(m, n - 1);
*/
/*@
  requires n == 0 || m == 0 || (n > 0 && m > 0);
  decreases n > 0 && m > 0 ? (integer)n + m : 0;
  assigns \nothing;
  ensures \result == (int32_t)fb_delannoy(n, m);
*/
int32_t dealnnoyNum(int32_t n, int32_t m) {
    if (m == 0 || n == 0) {
        return 1;
    }
    int32_t first = dealnnoyNum(java_sub(m, 1), n);
    int32_t second = dealnnoyNum(java_sub(m, 1), java_sub(n, 1));
    int32_t third = dealnnoyNum(m, java_sub(n, 1));
    return java_add(java_add(first, second), third);
}
